### PR TITLE
Replace CLI Search Expressions

### LINKED ISSUE #831: Replace usage of PyPEG2 with lark-parser
PyPEG2 is used for parsing search expressions, but hasn't been updated for years. We have had problems setting up Datacube environments when the PyPEG2 package has strange pinned dependencies.

Lark-parser is also much nicer to use.

### PR DESCRIPTION
**Pull request by omad**
### Reason for this pull request

I would like to replace the _search expressions_ used in several datacube command line tools with a more manageable and understandable alternative.

The existing search expressions have confused a few new users, they have grown since initially implemented, but haven't had old options removed. The implementation is also based on PyPEG2, a no longer supported or updated parsing library. Due to it's rarity, this dependency has created several dependency graph issues with datacube in conda-forge.

### Proposed changes

- Remove the PyPEG2 based expression parsing code, and the dependency on PyPEG2
- Replace with a simpler and more strictly defined implementation based on [lark parser](https://lark-parser.readthedocs.io/)
- Remove the old greater-than and less-than syntax. They were awkward to type, required shell escaping, and seemed to suggest that open ended queries were possible.
- Remove the `in range()` syntax, which also required shell escaping, and wasn't clear whether it supported single or multiple dates or only numbers.

<details><summary>Available Search Expressions</summary>

```
  Select datasets using [EXPRESSIONS] to filter by date, product type,
  spatial extents or other searchable fields.

      FIELD = VALUE
      FIELD in DATE-RANGE
      FIELD in [START, END]

  DATE-RANGE is one of YYYY, YYYY-MM or YYYY-MM-DD
  START and END can be either numbers or dates

  FIELD: x, y, lat, lon, time, product, ...

  eg. 'time in [1996-01-01, 1996-12-31]'
      'time in 1996'
      'lon in [130, 140]' 'lat in [-40, -30]'
      product=ls5_nbar_albers
```

</details>

Thanks @uchchwhash for a lot of the hard work!

 - [x] Closes #831
 - [x] Tests updated and passing.
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: datacube/scripts/dataset.py [hunk 1]
```diff
@@ -486,3 +486,14 @@ def within_tolerance(dataset):
...
+
+@dataset_cmd.command('delete', help="Delete datasets")
+@click.argument('ids', nargs=-1)
+@ui.pass_index()
+def dataset_delete(index, ids):
+    """
+    Delete Dataset(s) from the database
+
+    Easy if they have nothing derived from them.
+    """
```
Comment (jeremyh) [inline datacube/scripts/dataset.py]: Accidental inclusion? (empty stub methods)

**Review by jeremyh — DISMISSED** [review]
Nice, much simpler! :+1:

File: datacube/scripts/dataset.py (hunk 1 above)
Comment (omad) [reply datacube/scripts/dataset.py]: Whoops, thanks, yep. I'll clean it up. :)

File: datacube/scripts/product.py [hunk 2]
```diff
@@ -220,3 +220,17 @@ def show_product(dc, product_name, output_format):
...
+@ui.pass_index()
+def delete_product(index, product_name):
+    """
+    Delete a product from an ODC database.
+
+    If the product has no Datasets, no further confirmation is required.
+
+    If the product does have Datasets, suggest how to delete the datasets.
+
+    """
```
Comment (Kirill888) [inline datacube/scripts/product.py]: more accidental stub commits?

**Review by Kirill888 — APPROVED** [review]

**Pull request merged**